dialer.go

```diff
@@ -68,175 +81,177 @@ func NewDialer(opts *DialerOpts) Dialer {
+	log.Debugf("Initializing Dialer with   windowSize: %v   maxPadding: %v   liveConns: %v  maxStreamsPerConn: %v   pingInterval: %v   cipher: %v",
 		opts.WindowSize,
 		opts.MaxPadding,
-		opts.MaxLiveConns,
+		opts.LiveConns,
 		opts.MaxStreamsPerConn,
 		opts.PingInterval,
 		opts.Cipher)
-	liveSessions := make(chan sessionIntf, opts.MaxLiveConns)
-	liveSessions <- nullSession{}
-	return &dialer{
-		windowSize:            opts.WindowSize,
-		maxPadding:            opts.MaxPadding,
-		maxStreamsPerConn:     opts.MaxStreamsPerConn,
-		maxLiveConns:          opts.MaxLiveConns,
-		idleInterval:          opts.IdleInterval,
-		pingInterval:          opts.PingInterval,
-		redialSessionInterval: opts.RedialSessionInterval,
-		pool:            opts.Pool,
-		cipherCode:      opts.Cipher,
-		serverPublicKey: opts.ServerPublicKey,
-		liveSessions:    liveSessions,
-		numLive:         1, // the nullSession
-		emaRTT:          ema.NewDuration(0, 0.5),
+	var lc ClientLifecycleListener
+	if opts.Lifecycle != nil {
+		lc = opts.Lifecycle
+	} else {
+		lc = NoopClientLifecycleListener()
 	}
-}
 
-type dialer struct {
-	windowSize            int
-	maxPadding            int
-	maxLiveConns          int
-	maxStreamsPerConn     uint16
-	idleInterval          time.Duration
-	pingInterval          time.Duration
-	redialSessionInterval time.Duration
-	pool                  BufferPool
-	cipherCode            Cipher
-	serverPublicKey       *rsa.PublicKey
-	muNumLivePending      sync.Mutex
-	numLive               int
-	numPending            int
-	liveSessions          chan sessionIntf
-	emaRTT                *ema.EMA
-}
+	// We allow more than LiveConns here to accommodate the rare case that individual sessions
+	// fill up with streams, and we need more.
+	maxConns := opts.LiveConns * 2
+	d := &dialer{
+		windowSize:        opts.WindowSize,
+		maxPadding:        opts.MaxPadding,
+		maxStreamsPerConn: opts.MaxStreamsPerConn,
+		pingInterval:      opts.PingInterval,
+		pool:              opts.Pool,
+		cipherCode:        opts.Cipher,
+		serverPublicKey:   opts.ServerPublicKey,
+		liveSessions:      make(chan sessionIntf, maxConns),
+		emaRTT:            ema.NewDuration(0, 0.5),
+		dial:              opts.Dial,
+		pendingSessions:   make(chan *sessionConfig, maxConns),
+		sessionRequests:   make(chan bool, 10),
+		lifecyle:          lc,
+		name:              opts.Name,
+	}
+	d.lifecyle.OnStart()
+	for i := 0; i < opts.LiveConns-1; i++ {
+		d.pendingSessions <- &sessionConfig{
+			name:         "background to " + d.name,
+			dialTimeout:  defaultDialTimeout,
+			sleepOnError: 2 * time.Second,
+		}
+	}
 
-func (d *dialer) Dial(dial DialFN) (net.Conn, error) {
-	return d.DialContext(context.Background(), dial)
+	// We create another background session with a shorter dial timeout to ensure liveness in the case of network
+	// disruptions.
+	d.pendingSessions <- &sessionConfig{
+		name:         "liveness to " + d.name,
+		dialTimeout:  5 * time.Second,
```

Comment:
Would be good to make this configurable. Also, since we're doing these in the background anyway, I think a bigger timeout might be good for high congestion periods.

Reply:
Oh never mind, I see that the main one has a much longer dial timeout. Nice!